### PR TITLE
Implemented application links logic

### PR DESCRIPTION
**Pull request by RiccardoM**
## Description
This PR implements the application links logic. Tests till come in another PR.

Works towards #472

## Checklist
- [x] Targeted PR against correct branch.
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Wrote unit tests.
- [ ] Wrote integration tests (simulation & CLI). 
- [ ] Updated the documentation. 
- [ ] Added an entry to the `CHANGELOG.md` file.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.

### CODE REVIEW

File: x/profiles/client/cli/cli_app_links.go [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+	"github.com/desmos-labs/desmos/x/profiles/types"
+)
+
+const (
+	flagPacketTimeoutHeight    = "packet-timeout-height"
+	flagPacketTimeoutTimestamp = "packet-timeout-timestamp"
+	flagAbsoluteTimeouts       = "absolute-timeouts"
+)
+
+// GetCmdLinkApplication returns the command to create a NewMsgTransfer transaction
```
Comment (bragaz) [inline x/profiles/client/cli/cli_app_links.go:26]:
```suggestion
// GetCmdLinkApplication returns the command to create a NewMsgLinkApplication transaction
```

File: x/profiles/keeper/keeper_app_links.go [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+	link, err := k.GetApplicationLink(ctx, clientRequest.User, clientRequest.Application, clientRequest.Username)
+	if err != nil {
+		return "", types.ApplicationLink{}, err
+	}
+
+	return clientRequest.User, link, nil
+}
+
+func (k Keeper) DeleteApplicationLink(ctx sdk.Context, user string, application, username string) error {
```
Comment (bragaz) [inline x/profiles/keeper/keeper_app_links.go:100]: missing docs for this

File: x/profiles/keeper/msg_server_app_link.go [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+import (
+	"context"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/desmos-labs/desmos/x/profiles/types"
+)
+
+// LinkApplication defines a rpc handler method for MsgConnectProfile.
```
Comment (bragaz) [inline x/profiles/keeper/msg_server_app_link.go:11]:
```suggestion
// LinkApplication defines a rpc handler method for MsgLinkApplication.
```

File: x/profiles/keeper/keeper_app_links.go [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+	if err != nil {
+		return err
+	}
+
+	// Delete the client request
+	store.Delete(types.ApplicationLinkClientIDKey(link.OracleRequest.ClientID))
+
+	// Delete the link object
+	store.Delete(types.ApplicationLinkKey(user, application, username))
+
```
Comment (bragaz) [inline x/profiles/keeper/keeper_app_links.go:114]: Shouldn't we emit an event here also? Like in the `SaveApplicationLink` above

Comment (RiccardoM) [reply x/profiles/keeper/keeper_app_links.go:116]: Currently the event is emitted inside the `msg_server`. We should move it to the keeper when handling #482

Comment (bragaz) [reply x/profiles/keeper/keeper_app_links.go:116]: Cool 👍

**Review by bragaz — COMMENTED** [review]
Some failing checks. I will re-review this when fixed.
Apart from this, looks great 👍

**Review by bragaz — COMMENTED** [review]
review-2: LGTM

File: x/profiles/module_ibc.go [hunk 5]
```diff
@@ -179,8 +279,28 @@ func (am AppModule) OnAcknowledgementPacket(
...
 	ctx sdk.Context,
-	modulePacket channeltypes.Packet,
+	packet channeltypes.Packet,
 ) (*sdk.Result, error) {
+	var data oracletypes.OracleRequestPacketData
+	if err := types.AminoCdc.UnmarshalJSON(packet.GetData(), &data); err != nil {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrUnknownRequest,
+			"cannot unmarshal oracle request packet data: %s", err.Error())
+	}
+	// refund tokens
```
Comment (dadamu) [inline x/profiles/module_ibc.go:289]:
```suggestion

```

File: x/profiles/module_ibc.go [hunk 6]
```diff
@@ -179,8 +279,28 @@ func (am AppModule) OnAcknowledgementPacket(
 // OnTimeoutPacket implements the IBCModule interface
 func (am AppModule) OnTimeoutPacket(
 	ctx sdk.Context,
-	modulePacket channeltypes.Packet,
+	packet channeltypes.Packet,
 ) (*sdk.Result, error) {
+	var data oracletypes.OracleRequestPacketData
+	if err := types.AminoCdc.UnmarshalJSON(packet.GetData(), &data); err != nil {
```
Comment (dadamu) [inline x/profiles/module_ibc.go:285]:
```suggestion
	if err := types.ProtoCdc.UnmarshalJSON(packet.GetData(), &data); err != nil {
```
The codec for packet should be `ProtoCdc`.

File: x/profiles/module_ibc.go [hunk 7]
```diff
@@ -149,28 +184,93 @@ func (am AppModule) OnRecvPacket(
...
 
 // OnAcknowledgementPacket implements the IBCModule interface
 func (am AppModule) OnAcknowledgementPacket(
 	ctx sdk.Context,
-	modulePacket channeltypes.Packet,
+	packet channeltypes.Packet,
 	acknowledgement []byte,
 ) (*sdk.Result, error) {
+	var ack channeltypes.Acknowledgement
+	if err := types.AminoCdc.UnmarshalJSON(acknowledgement, &ack); err != nil {
```
Comment (dadamu) [inline x/profiles/module_ibc.go:233]:
```suggestion
	if err := types.ProtoCdc.UnmarshalJSON(acknowledgement, &ack); err != nil {
```
The codec for packet should be `ProtoCdc`.

File: x/profiles/module_ibc.go [hunk 8]
```diff
@@ -149,28 +184,93 @@ func (am AppModule) OnRecvPacket(
...
+
+	switch resp := ack.Response.(type) {
+	case *channeltypes.Acknowledgement_Result:
+		ctx.EventManager().EmitEvent(
+			sdk.NewEvent(
+				types.EventTypePacket,
+				sdk.NewAttribute(types.AttributeKeyAckSuccess, string(resp.Result)),
+			),
+		)
+	case *channeltypes.Acknowledgement_Error:
```
Comment (dadamu) [inline x/profiles/module_ibc.go:265]: This case can not be reached cause the error acknowledgment would be returned by `OnOracleRequestAcknowledgementPacket`.
The error should not be returned early if  we would like to emit event with error attribute here.

File: x/profiles/module_ibc.go [hunk 9]
```diff
@@ -149,28 +184,93 @@ func (am AppModule) OnRecvPacket(
...
 	}
 
+	ctx.EventManager().EmitEvent(
+		sdk.NewEvent(
+			types.EventTypePacket,
+			sdk.NewAttribute(sdk.AttributeKeyModule, types.ModuleName),
+			sdk.NewAttribute(types.AttributeKeyClientID, data.ClientID),
+			sdk.NewAttribute(types.AttributeKeyRequestID, fmt.Sprintf("%d", data.RequestID)),
+			sdk.NewAttribute(types.AttributeKeyResolveStatus, data.ResolveStatus.String()),
+			sdk.NewAttribute(types.AttributeKeyAckSuccess, fmt.Sprintf("%t", err != nil)),
```
Comment (dadamu) [inline x/profiles/module_ibc.go:218]:
```suggestion
			sdk.NewAttribute(types.AttributeKeyAckSuccess, fmt.Sprintf("%t", true)),
```

File: x/profiles/keeper/relay_app_links.go [hunk 10]
```diff
@@ -0,0 +1,268 @@
...
+	sequence, found := k.channelKeeper.GetNextSequenceSend(ctx, sourcePort, sourceChannel)
+	if !found {
+		return sdkerrors.Wrapf(
+			channeltypes.ErrSequenceSendNotFound,
+			"source port: %s, source channel: %s", sourcePort, sourceChannel,
+		)
+	}
+
+	// Begin createOutgoingPacket logic
+	// See spec for this logic: https://github.com/cosmos/ics/tree/master/spec/ics-020-fungible-token-transfer#packet-relay
```
Comment (dadamu) [inline x/profiles/keeper/relay_app_links.go:82]:
```suggestion
	// See spec for this logic: https://github.com/cosmos/ibc/tree/master/spec/app/ics-020-fungible-token-transfer#packet-relay
```

Comment (dadamu) [inline x/profiles/keeper/relay_app_links.go:82]: The spec is for `fungible-token-transfer`. I think it should be removed.

File: x/profiles/keeper/keeper_app_links.go [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+	return clientRequest.User, link, nil
+}
+
+// DeleteApplicationLink removes the application link associated to the given user,
+// for the given application and username
+func (k Keeper) DeleteApplicationLink(ctx sdk.Context, user string, application, username string) error {
+	store := ctx.KVStore(k.storeKey)
+
+	// Get the link to obtain che client id
```
Comment (dadamu) [inline x/profiles/keeper/keeper_app_links.go:105]:
```suggestion
	// Get the link to obtain the client id
```

File: x/profiles/types/msgs_app_links.go [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	clienttypes "github.com/cosmos/cosmos-sdk/x/ibc/core/02-client/types"
+	host "github.com/cosmos/cosmos-sdk/x/ibc/core/24-host"
+)
+
+const (
+	TypeMsgLinkApplication   = "link-application"
+	TypeMsgUnlinkApplication = "unlink-application"
```
Comment (dadamu) [inline x/profiles/types/msgs_app_links.go:14]: I think they should be put into keys.go, renamed into `ActionLinkApplication` and `ActionUnlinkApplication` with value of snake case.

**Review by dadamu — CHANGES_REQUESTED** [review]
It almost looks great. Just some failing checks.

File: x/profiles/module_ibc.go (hunk 8 above)
Comment (RiccardoM) [reply x/profiles/module_ibc.go:265]: This is not actually true though. Inside `OnOracleRequestAcknowledgementPacket`, the error is returned only if the storing of the updated link errors: 

```golang
return k.SaveApplicationLink(ctx, user, link)
```

Inside such method, the acknowledgement error is only used to update the link state properly:

```golang
case *channeltypes.Acknowledgement_Error:
	// The acknowledgment failed on the receiving chain.
	// Update the state to ERROR and the result to an error one
	link.State = types.AppLinkStateVerificationError
	link.Result = types.NewErrorResult(res.Error)
```

**Comment by RiccardoM** [comment]
Checks will be fixed with the tests PR

File: x/profiles/module_ibc.go (hunk 8 above)
Comment (dadamu) [reply x/profiles/module_ibc.go:268]: @RiccardoM You are right. I missed that thing.

File: x/profiles/module_ibc.go [hunk 13]
```diff
@@ -117,29 +119,62 @@ func (am AppModule) OnChanCloseConfirm(
...
+	ack, err = am.handleLinkChainAccountPacketData(ctx, packet)
+	if types.ErrInvalidPacketData.Is(err) {
+		// Try handling the oracle request packet data
+		ack, err = am.handleOracleRequestPacketData(ctx, packet)
+	}
+
+	// If packet data is still invalid, return an error
+	if types.ErrInvalidPacketData.Is(err) {
+		return nil, nil, err
+	}
```
Comment (dadamu) [inline x/profiles/module_ibc.go:137]: Cant it be switch case to handle packet data?

Say, 
```
switch packet.(type) {
case *types.LinkChainAccountPacketData:
  am.handleLinkChainAccountPacketData(ctx, packet)
case *oracletypes.OracleResponsePacketData:
  am.handleOracleRequestPacketData(ctx, packet)
default:
  return nil, nil, err
}
```

**Review by dadamu — COMMENTED** [review]
Just have a last concern about packet in OnRecvPacket.
Apart from it, it looks great to me

File: x/profiles/module_ibc.go (hunk 13 above)
Comment (RiccardoM) [reply x/profiles/module_ibc.go:137]: Unfortunately, this is not possible since `Packet` is not an interface

Comment (dadamu) [reply x/profiles/module_ibc.go:137]: @RiccardoM I see. It makes sense to me now.

**Review by dadamu — APPROVED** [review]
LGTM

File: x/profiles/module_ibc.go [hunk 14]
```diff
@@ -114,32 +154,82 @@ func (am AppModule) OnChanCloseConfirm(
...
+	packet channeltypes.Packet,
 ) (*sdk.Result, []byte, error) {
-	var packetData types.LinkChainAccountPacketData
-	if err := types.ProtoCdc.UnmarshalJSON(modulePacket.GetData(), &packetData); err != nil {
-		return nil, nil, sdkerrors.Wrapf(sdkerrors.ErrUnknownRequest, "cannot unmarshal packet data: %s", err.Error())
+	var ack channeltypes.Acknowledgement
+	var err error
+
+	// Try handling the chain link packet data
+	ack, err = am.HandlePacket(ctx, packet, handleOracleRequestPacketData, handleLinkChainAccountPacketData)
```
Comment (RiccardoM) [inline x/profiles/module_ibc.go:168]: @dadamu I'm going to merge the PR, but just as a reference I've changed how multiple packet data types are handled. You can see the new logic here

**Pull request merged**